File: src/server/controllers/export-character.ts
import chalk from 'chalk';
import { createHash, randomUUID } from 'crypto';
import express from 'express';
import fsExtra from 'fs-extra';
import _ from 'lodash';
import path from 'path';
import z from 'zod';

import {
  CharacterExporter, CharacterSchema, LocalRefSchema, LocalRefValueSchema,
} from '@/lib/converter/character';
import { Config, getDefaultConfig } from '@/lib/global-config';
import { stableStringify, waitUntil } from '@/lib/utils';
import { wowExportClient } from '@/lib/wowexport-client/wowexport-client';
import { Job, JobQueue, QueueConfig } from '@/server/utils/job-queue';

import {
  isSharedHosting, outputDir, outputDirBrowse, serverDeployTime,
} from '../config';
import { startupRequests } from './export-character.startup';

export const ExporCharacterRequestSchema = z.object({
  character: CharacterSchema,
  outputFileName: LocalRefValueSchema,
  optimization: z.object({
    sortSequences: z.boolean().optional(),
    removeUnusedVertices: z.boolean().optional(),
    removeUnusedNodes: z.boolean().optional(),
    removeUnusedMaterialsTextures: z.boolean().optional(),
    maxTextureSize: z.enum(['256', '512', '1024']).optional(),
  }),
  format: z.enum(['mdx', 'mdl']),
  formatVersion: z.enum(['800', '1000']).optional(),
  isBrowse: z.boolean().optional(),
});

export type ExportCharacterRequest = z.infer<typeof ExporCharacterRequestSchema>;

export type ExportCharacterResponse = {
  exportedModels: { path: string, size: number }[]
  exportedTextures: { path: string, size: number }[]
  modelStats: {
    formatVersion: number;
    vertices: number;
    faces: number;
    globalSequences: number;
    sequences: number;
    geosets: number;
    geosetAnims: number;
    materials: number;
    textures: number;
    textureAnims: number;
    bones: number;
    lights: number;
    ribbonEmitters: number;
    particles: number;
    attachments: number;
    eventObjects: number;
    helpers: number;
    collisionShapes: number;
    cameras: number;
  }
  outputDirectory?: string
  versionId: string
}

const queueConfig: QueueConfig<ExportCharacterRequest, ExportCharacterResponse> = {
  concurrency: 1,
  maxPendingJobs: 100,
  jobTTL: 5 * 60 * 1000,
  jobTimeout: 60 * 1000 + (isSharedHosting ? 60 * 1000 : 0),
};

type ExportCharacterJob = Job<ExportCharacterRequest, ExportCharacterResponse>;

let ceConfig: Config;

export function getCeConfig() {
  return ceConfig;
}

let logs: string[] = [];
const originalConsoleLog = console.log;
console.log = (...args) => {
  logs.push(args.join(' '));
  if (logs.length > 2) logs.shift();
  originalConsoleLog(...args);
};

export async function ControllerExportCharacter(router: express.Router) {
  await waitUntil(() => wowExportClient.isReady);
  ceConfig = await getDefaultConfig();
  if (isSharedHosting) {
    console.log('Shared hosting mode enabled');
  }

  /** Core export logic, extracted into its own function so the queue worker can reuse it */
  async function handleExport(job: ExportCharacterJob) {
    const start = performance.now();
    const request = job.request;
    const ce = new CharacterExporter({
      ...ceConfig,
      maxTextureSize: request.optimization.maxTextureSize
        ? parseInt(request.optimization.maxTextureSize, 10)
        : undefined,
    });
    console.log(`Start exporting ${request.outputFileName}: ${chalk.gray(JSON.stringify(request, null, 2))}`);
    logs = [];

    await wowExportClient.syncConfig();
    await ce.exportCharacter(request.character, request.outputFileName);

    ce.models.forEach(([mdl]) => {
      if (request.formatVersion === '800') {
        mdl.modify.convertToSd800();
        mdl.materials.forEach((m) => {
          m.layers.forEach((l) => {
            l.unshaded = true;
          });
        });
      }
      if (request.optimization?.sortSequences) {
        mdl.modify.sortSequences();
      }
      if (request.optimization?.removeUnusedVertices) {
        mdl.modify.removeUnusedVertices();
      }
      const particlesDensity = request.character.particlesDensity;
      if (particlesDensity != null) {
        if (particlesDensity > 0 && particlesDensity !== 1) {
          mdl.modify.scaleParticlesDensity(particlesDensity);
        } else if (particlesDensity === 0) {
          mdl.particleEmitter2s = [];
        }
      }
      if (request.optimization?.removeUnusedNodes) {
        mdl.modify.removeUnusedNodes();
      }
      if (request.optimization?.removeUnusedMaterialsTextures) {
        mdl.modify.removeUnusedMaterialsTextures();
      }
      mdl.modify.optimizeKeyFrames();
      mdl.sync();
    });

    const targetDir = request.isBrowse ? outputDirBrowse : outputDir;
    const modelPaths = ce.writeAllModels(targetDir, request.format);
    const texturePaths = await ce.writeAllTextures(targetDir);

    const exportedModels = modelPaths.map((modelPath) => ({
      path: path.relative(targetDir, `${modelPath}.${request.format}`),
      size: fsExtra.statSync(`${modelPath}.${request.format}`).size,
    }));
    const exportedTextures = texturePaths.map((texturePath) => ({
      path: path.relative(targetDir, texturePath),
      size: fsExtra.statSync(texturePath).size,
    }));
    exportedTextures.sort((a, b) => a.path.localeCompare(b.path));

    // Return the list of exported assets to the caller – zipping happens on-demand via the download API
    const resp: ExportCharacterResponse = {
      exportedModels,
      exportedTextures,
      outputDirectory: !isSharedHosting ? path.resolve(targetDir) : undefined,
      versionId: job.id,
      modelStats: {
        formatVersion: request.formatVersion === '800' ? 800 : 1000,
        globalSequences: ce.models.reduce((acc, [mdl]) => acc + mdl.globalSequences.length, 0),
        sequences: ce.models.reduce((acc, [mdl]) => acc + mdl.sequences.length, 0),
        geosets: ce.models.reduce((acc, [mdl]) => acc + mdl.geosets.length, 0),
        geosetAnims: ce.models.reduce((acc, [mdl]) => acc + mdl.geosetAnims.length, 0),
        materials: ce.models.reduce((acc, [mdl]) => acc + mdl.materials.length, 0),
        textures: ce.models.reduce((acc, [mdl]) => acc + mdl.textures.length, 0),
        textureAnims: ce.models.reduce((acc, [mdl]) => acc + mdl.textureAnims.length, 0),
        bones: ce.models.reduce((acc, [mdl]) => acc + mdl.bones.length, 0),
        lights: ce.models.reduce((acc, [mdl]) => acc + mdl.lights.length, 0),
        ribbonEmitters: ce.models.reduce((acc, [mdl]) => acc + mdl.ribbonEmitters.length, 0),
        particles: ce.models.reduce((acc, [mdl]) => acc + mdl.particleEmitter2s.length, 0),
        attachments: ce.models.reduce((acc, [mdl]) => acc + mdl.attachments.length, 0),
        eventObjects: ce.models.reduce((acc, [mdl]) => acc + mdl.eventObjects.length, 0),
        helpers: ce.models.reduce((acc, [mdl]) => acc + mdl.helpers.length, 0),
        collisionShapes: ce.models.reduce((acc, [mdl]) => acc + mdl.collisionShapes.length, 0),
        cameras: ce.models.reduce((acc, [mdl]) => acc + mdl.cameras.length, 0),
        vertices: ce.models.reduce((acc, [mdl]) => acc + mdl.geosets.reduce((acc, g) => acc + g.vertices.length, 0), 0),
        faces: ce.models.reduce((acc, [mdl]) => acc + mdl.geosets.reduce((acc, g) => acc + g.faces.length, 0), 0),
      },
    };

    console.log(
      'Job finished',
      `${chalk.yellow(`${((performance.now() - start) / 1000).toFixed(2)}s`)}`,
      chalk.gray(JSON.stringify(_.omit(resp, 'exportedTextures', 'modelStats'), null, 2)),
    );

    return resp;
  }

  const jobQueue = new JobQueue<ExportCharacterRequest, ExportCharacterResponse>(
    {
      ...queueConfig,
      jobCompletedCallback: () => {
        fsExtra.writeFileSync('recent-exports.json', JSON.stringify(jobQueue.recentCompletedJobs, null, 2));
      },
    },
    (job) => handleExport(job),
  );

  // Load recent exports from file so that it survives server restart
  try {
    const recentExports = JSON.parse(fsExtra.readFileSync('recent-exports.json', 'utf8')) as ExportCharacterJob[];
    jobQueue.recentCompletedJobs = recentExports;
  } catch (err) {
    // Ignore
  }

  router.get('/export/character/recent', (req, res) => {
    res.json(jobQueue.recentCompletedJobs);
  });

  router.post('/export/character', (req, res) => {
    if (!wowExportClient.isReady) {
      return res.status(500).json({ error: 'wow.export RCP server is not ready' });
    }

    try {
      const parsedRequest = ExporCharacterRequestSchema.parse(req.body);

      // Use a unique version suffix for the model file names to prevent collisions in multi-tenant scenarios
      // while keeping all assets in the shared exported-assets folder (avoids duplicating textures).
      // Deterministic version hash based on request content + server epoch when in shared hosting mode.
      const versionId = isSharedHosting || parsedRequest.isBrowse
        ? createHash('md5').update(stableStringify(parsedRequest) + serverDeployTime).digest('hex')
        : undefined;
      if (versionId) {
        parsedRequest.outputFileName = `${parsedRequest.outputFileName}__${versionId}`;
      }

      if (versionId && jobQueue.getJob(versionId)) {
        return res.json(jobQueue.getJobStatus(versionId));
      }

      const job: ExportCharacterJob = {
        id: versionId ?? randomUUID(),
        request: parsedRequest,
        status: 'pending',
        submittedAt: Date.now(),
      };

      jobQueue.addJob(job);
      console.log(chalk.blue(req.method), req.path, chalk.gray(`Queued job ${job.id}`));

      setImmediate(() => {
        res.json(jobQueue.getJobStatus(job.id));
      });
      return res;
    } catch (err) {
      return res.status(400).json({ error: err instanceof Error ? err.message : String(err) });
    }
  });

  router.get('/export/character/status/:jobId', (req, res) => {
    const { jobId } = req.params;
    const job = jobQueue.getJob(jobId);

    if (!job) {
      return res.status(404).json({ error: 'Export request not found' });
    }

    const status = jobQueue.getJobStatus(jobId);

    return res.json({
      ...status,
      logs: status?.status === 'processing' ? logs : undefined,
    });
  });

  router.post('/export/character/clean', (req, res) => {
    if (outputDir !== 'exported-assets') {
      return res.status(500).json({ error: 'Cannot clean exported assets because output directory is not "exported-assets"' });
    }
    if (isSharedHosting) {
      return res.status(400).json({ error: 'Cannot clean exported assets because server is in shared hosting mode' });
    }

    fsExtra.removeSync('./recent-exports.json');
    jobQueue.recentCompletedJobs = [];
    console.log(`Removed ${path.resolve('./recent-exports.json')}`);
    fsExtra.emptyDirSync(outputDir);
    console.log(`Cleared ${path.resolve(outputDir)}`);
    fsExtra.emptyDirSync(outputDirBrowse);
    console.log(`Cleared ${path.resolve(outputDirBrowse)}`);
    wowExportClient.clearCacheFiles();
    console.log(`Cleared ${path.resolve('.cache')}`);
    return res.json({ message: 'Exported assets cleaned' });
  });

  // Serve exported assets. When running on shared hosting enable HTTP caching to reduce bandwidth
  const exportStatic = (dir: string) => (isSharedHosting
    ? express.static(dir, {
      maxAge: '1h',
      setHeaders: (res) => {
        res.setHeader('Cache-Control', 'public, max-age=3600');
      },
    })
    : express.static(dir));
  router.use('/browse-assets', exportStatic(outputDirBrowse));
  router.use('/assets', exportStatic(outputDir));
  fsExtra.emptyDirSync(outputDirBrowse);

  let jobs: ExportCharacterJob[] = [];
  if (isSharedHosting) {
    jobs = startupRequests.map((request) => {
      const job: ExportCharacterJob = {
        id: randomUUID(),
        request,
        status: 'pending',
        submittedAt: Date.now(),
        isDemo: true,
      };
      jobQueue.addJob(job);
      return job;
    });
  }

  router.get('/export/character/demos', (req, res) => {
    res.json(jobs.map((job) => jobQueue.getJob(job.id)).filter((job) => job?.status === 'done'));
  });

  router.get('/export/character/check-local-file', (req, res) => {
    try {
      const { localPath } = req.query as { localPath: string };
      const parsed = LocalRefSchema.safeParse({ type: 'local', value: localPath });
      if (!parsed.success) {
        return res.json({
          ok: false,
          similarFiles: [],
        });
      }

      let baseModelPath = path.resolve(ceConfig.wowExportAssetDir, localPath);
      // prevent traversal attacks
      if (!baseModelPath.startsWith(ceConfig.wowExportAssetDir)) {
        return res.json({
          ok: false,
          similarFiles: [],
        });
      }
      if (!baseModelPath.endsWith('.obj')) {
        baseModelPath += '.obj';
      }
      const searchPrefix = baseModelPath.replace('.obj', '');
      const dirName = path.dirname(baseModelPath);
      if (!fsExtra.existsSync(dirName)) {
        return res.json({
          ok: false,
          similarFiles: [],
        });
      }
      const allFiles = fsExtra.readdirSync(dirName).map((file) => path.join(dirName, file));
      const similarFiles = allFiles
        .filter((file) => file.startsWith(searchPrefix) && file.endsWith('.obj') && !file.endsWith('.phys.obj'))
        .map((file) => path.relative(ceConfig.wowExportAssetDir, file));

      return res.json({
        ok: fsExtra.existsSync(baseModelPath),
        similarFiles,
      });
    } catch (err) {
      return res.json({
        ok: false,
        similarFiles: [],
      });
    }
  });
}

File: golang/internal/server/api/export_character.go
package api

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"errors"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/pqhuy98/wow-converter/internal/ansi"
	"github.com/pqhuy98/wow-converter/internal/config"
	"github.com/pqhuy98/wow-converter/internal/converter/character"
	"github.com/pqhuy98/wow-converter/internal/converter/common"
	"github.com/pqhuy98/wow-converter/internal/server/exportlog"
	"github.com/pqhuy98/wow-converter/internal/server/pathsafe"
	"github.com/pqhuy98/wow-converter/internal/server/util"
	"github.com/pqhuy98/wow-converter/internal/stringsort"
	"github.com/pqhuy98/wow-converter/internal/workspace"
	"github.com/pqhuy98/wow-converter/internal/wow/client"
)

type exportCharacterRequest struct {
	Character      character.Character `json:"character"`
	OutputFileName string              `json:"outputFileName"`
	Optimization   struct {
		SortSequences                 *bool  `json:"sortSequences"`
		AllMaterialsUnshaded          *bool  `json:"allMaterialsUnshaded"`
		RemoveUnusedVertices          *bool  `json:"removeUnusedVertices"`
		RemoveUnusedNodes             *bool  `json:"removeUnusedNodes"`
		RemoveUnusedMaterialsTextures *bool  `json:"removeUnusedMaterialsTextures"`
		MaxTextureSize                string `json:"maxTextureSize"`
	} `json:"optimization"`
	Format        string `json:"format"`
	FormatVersion string `json:"formatVersion"`
	IsBrowse      bool   `json:"isBrowse"`
	SkinID        string `json:"skinId"`
}

type exportCharacterResponse struct {
	ExportedModels   []exportAssetInfo `json:"exportedModels"`
	ExportedTextures []exportAssetInfo `json:"exportedTextures"`
	ModelStats       map[string]int    `json:"modelStats"`
	OutputDirectory  string            `json:"outputDirectory,omitempty"`
	VersionID        string            `json:"versionId"`
}

type exportAssetInfo struct {
	Path string `json:"path"`
	Size int64  `json:"size"`
}

type exportCharacterJobStatus struct {
	util.JobStatusView[exportCharacterResponse]
	Logs []string `json:"logs,omitempty"`
}

func registerExportCharacter(r Router, d *Deps) {
	exportlog.Install()

	timeout := 60 * time.Second
	if d.Config.IsSharedHosting {
		timeout += 60 * time.Second
	}

	var queue *util.JobQueue[exportCharacterRequest, exportCharacterResponse]
	queue = util.NewJobQueue(util.QueueConfig[exportCharacterRequest, exportCharacterResponse]{
		Concurrency:    1,
		MaxPendingJobs: 100,
		JobTTL:         5 * time.Minute,
		JobTimeout:     timeout,
		OnCompleted: func(_ *util.Job[exportCharacterRequest, exportCharacterResponse]) {
			_ = saveRecentExports(d.Config.RecentExports, queue.RecentCompletedJobs)
		},
	}, func(job *util.Job[exportCharacterRequest, exportCharacterResponse]) (exportCharacterResponse, error) {
		exportlog.Begin()
		defer exportlog.End()
		return runCharacterExport(job.Request, job.ID, d)
	})
	loadRecentExports(queue, d.Config.RecentExports)

	r.Get("/export/character/recent", func(w http.ResponseWriter, _ *http.Request) {
		sendJSON(w, http.StatusOK, queue.RecentCompletedJobs)
	})

	r.Post("/export/character", func(w http.ResponseWriter, req *http.Request) {
		if !d.IsDataServerReady(req.Context()) {
			sendError(w, http.StatusInternalServerError, "wow-data-server is not ready")
			return
		}
		rawBody, err := readJSONMap(req)
		if err != nil {
			if errors.Is(err, errRequestBodyTooLarge) {
				sendError(w, http.StatusRequestEntityTooLarge, "Request body too large")
				return
			}
			sendError(w, http.StatusBadRequest, "Invalid request body")
			return
		}
		body, issues := parseExportCharacterRequest(rawBody)
		if len(issues) > 0 {
			sendJSON(w, http.StatusBadRequest, map[string]any{
				"error":  "Invalid request body",
				"issues": issues,
			})
			return
		}

		var versionID *string
		if d.Config.IsSharedHosting || body.IsBrowse {
			hash := md5.Sum(append(mustJSON(body), []byte(d.Config.ServerDeployTime)...))
			v := hex.EncodeToString(hash[:])
			versionID = &v
			body.OutputFileName = body.OutputFileName + "__" + v
		}

		if versionID != nil {
			if existing := queue.GetJob(*versionID); existing != nil {
				sendJSON(w, http.StatusOK, queue.GetJobStatus(*versionID))
				return
			}
		}

		id := newJobID()
		if versionID != nil {
			id = *versionID
		}
		job := &util.Job[exportCharacterRequest, exportCharacterResponse]{
			ID: id, Request: body, Status: util.JobPending,
			SubmittedAt: time.Now().UnixMilli(), AddToRecent: !body.IsBrowse,
		}
		queue.AddJob(job)
		log.Printf("%s %s %s", ansi.Blue("POST"), "/export/character", ansi.Grayf("Queued job %s", job.ID))
		sendJSON(w, http.StatusOK, queue.GetJobStatus(job.ID))
	})

	r.Get("/export/character/status/{jobId}", func(w http.ResponseWriter, req *http.Request) {
		status := queue.GetJobStatus(chi.URLParam(req, "jobId"))
		if status == nil {
			sendError(w, http.StatusNotFound, "Export request not found")
			return
		}
		resp := exportCharacterJobStatus{JobStatusView: *status}
		if status.Status == util.JobProcessing {
			resp.Logs = exportlog.Snapshot()
		}
		sendJSON(w, http.StatusOK, resp)
	})

	r.Post("/export/character/clean", func(w http.ResponseWriter, _ *http.Request) {
		if d.Config.OutputDir != workspace.ResolveRepoPath(outputDirRel) {
			sendError(w, http.StatusInternalServerError, `Cannot clean exported assets because output directory is not "exported-assets"`)
			return
		}
		if d.Config.IsSharedHosting {
			sendError(w, http.StatusBadRequest, "Cannot clean exported assets because server is in shared hosting mode")
			return
		}
		_ = os.Remove(d.Config.RecentExports)
		queue.RecentCompletedJobs = emptyRecentExports()
		log.Printf("Removed %s", d.Config.RecentExports)
		_ = os.RemoveAll(d.Config.OutputDir)
		_ = os.MkdirAll(d.Config.OutputDir, 0o755)
		log.Printf("Cleared %s", d.Config.OutputDir)
		_ = os.RemoveAll(d.Config.OutputDirBrowse)
		_ = os.MkdirAll(d.Config.OutputDirBrowse, 0o755)
		log.Printf("Cleared %s", d.Config.OutputDirBrowse)
		sendJSON(w, http.StatusOK, map[string]string{"message": "Exported assets cleaned"})
	})

	var demoJobIDs []string
	if d.Config.IsSharedHosting {
		for _, demoReq := range StartupDemoRequests() {
			job := &util.Job[exportCharacterRequest, exportCharacterResponse]{
				ID: newJobID(), Request: demoReq, Status: util.JobPending,
				SubmittedAt: time.Now().UnixMilli(), AddToRecent: false, NoTimeout: true,
			}
			demoJobIDs = append(demoJobIDs, job.ID)
			queue.AddJob(job)
		}
	}

	r.Get("/export/character/demos", func(w http.ResponseWriter, _ *http.Request) {
		out := make([]*exportCharacterJobStatus, 0, len(demoJobIDs))
		for _, id := range demoJobIDs {
			if status := queue.GetJobStatus(id); status != nil && status.Status == util.JobDone {
				jobStatus := exportCharacterJobStatus{JobStatusView: *status}
				out = append(out, &jobStatus)
			}
		}
		sendJSON(w, http.StatusOK, out)
	})

	r.Get("/export/character/check-local-file", func(w http.ResponseWriter, req *http.Request) {
		localPath := req.URL.Query().Get("localPath")
		if localPath == "" {
			sendJSON(w, http.StatusOK, map[string]any{"ok": false, "similarFiles": []string{}})
			return
		}
		normalized := common.NormalizeLocalModelRef(localPath)
		if _, err := pathsafe.ResolveUnderBase(d.ExportAssetDir(), normalized); err != nil {
			sendJSON(w, http.StatusOK, map[string]any{"ok": false, "similarFiles": []string{}})
			return
		}
		ok, similar, err := character.ResolveLocalModelRef(d.Client, localPath)
		if err != nil {
			sendJSON(w, http.StatusOK, map[string]any{"ok": false, "similarFiles": []string{}})
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{"ok": ok, "similarFiles": similar})
	})

	registerExportStatic(r, d)
	if !d.Config.IsSharedHosting {
		_ = os.RemoveAll(d.Config.OutputDirBrowse)
		_ = os.MkdirAll(d.Config.OutputDirBrowse, 0o755)
	}
}

func registerExportStatic(r Router, d *Deps) {
	maxAge := "max-age=5"
	if d.Config.IsSharedHosting {
		maxAge = "max-age=3600"
	}
	assetsFS := http.StripPrefix("/api/assets/", http.FileServer(http.Dir(d.Config.OutputDir)))
	browseFS := http.StripPrefix("/api/browse-assets/", http.FileServer(http.Dir(d.Config.OutputDirBrowse)))
	r.Handle("/assets/*", cacheControlHandler(maxAge, assetsFS))
	r.Handle("/browse-assets/*", cacheControlHandler(maxAge, browseFS))
}

func cacheControlHandler(cacheControl string, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		w.Header().Set("Cache-Control", "public, "+cacheControl)
		next.ServeHTTP(w, req)
	})
}

func mustJSON(v any) []byte {
	b, _ := json.Marshal(v)
	return b
}

func loadRecentExports(queue *util.JobQueue[exportCharacterRequest, exportCharacterResponse], path string) {
	queue.RecentCompletedJobs = emptyRecentExports()
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}
	_ = json.Unmarshal(data, &queue.RecentCompletedJobs)
	if queue.RecentCompletedJobs == nil {
		queue.RecentCompletedJobs = emptyRecentExports()
	}
}

func emptyRecentExports() []*util.Job[exportCharacterRequest, exportCharacterResponse] {
	return []*util.Job[exportCharacterRequest, exportCharacterResponse]{}
}

func saveRecentExports(path string, jobs []*util.Job[exportCharacterRequest, exportCharacterResponse]) error {
	data, err := json.MarshalIndent(jobs, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

func optimizationBool(v *bool, defaultVal bool) bool {
	if v == nil {
		return defaultVal
	}
	return *v
}

func runCharacterExport(req exportCharacterRequest, jobID string, d *Deps) (exportCharacterResponse, error) {
	startedAt := time.Now()
	requestJSON, _ := json.MarshalIndent(req, "", "  ")
	log.Printf("Start exporting %s: %s", req.OutputFileName, ansi.Gray(string(requestJSON)))

	cfg := config.DefaultConfig()
	cfg.MDX = req.Format == "mdx"
	if req.Optimization.MaxTextureSize != "" {
		switch req.Optimization.MaxTextureSize {
		case "256":
			cfg.MaxTextureSize = 256
		case "512":
			cfg.MaxTextureSize = 512
		case "1024":
			cfg.MaxTextureSize = 1024
		}
	}

	outDir := d.Config.OutputDir
	if req.IsBrowse {
		outDir = d.Config.OutputDirBrowse
	}
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		return exportCharacterResponse{}, err
	}

	exporter := character.NewCharacterExporter(cfg, d.Client)
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Minute)
	defer cancel()

	if err := client.SyncConfig(ctx, d.Client); err != nil {
		log.Printf("syncConfig: %v", err)
	}

	_, err := exporter.ExportCharacter(ctx, req.Character, req.OutputFileName, character.ExportOptions{
		LocalModelSkinID: req.SkinID,
	})
	if err != nil {
		return exportCharacterResponse{}, err
	}
	exporter.OptimizeModelsTextures(character.ExportOptimization{
		SortSequences:                 optimizationBool(req.Optimization.SortSequences, true),
		RemoveUnusedVertices:          optimizationBool(req.Optimization.RemoveUnusedVertices, true),
		RemoveUnusedNodes:             optimizationBool(req.Optimization.RemoveUnusedNodes, true),
		RemoveUnusedMaterialsTextures: optimizationBool(req.Optimization.RemoveUnusedMaterialsTextures, true),
		FormatVersion:                 req.FormatVersion,
		AllMaterialsUnshaded:          req.Optimization.AllMaterialsUnshaded != nil && *req.Optimization.AllMaterialsUnshaded,
		ParticlesDensity:              req.Character.ParticlesDensity,
	})

	var texturePaths, modelPaths []string
	var texErr, modelErr error
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		texturePaths, texErr = exporter.WriteAllTextures(outDir)
	}()
	go func() {
		defer wg.Done()
		modelPaths, modelErr = exporter.WriteAllModels(outDir, req.Format)
	}()
	wg.Wait()
	if texErr != nil {
		return exportCharacterResponse{}, texErr
	}
	if modelErr != nil {
		return exportCharacterResponse{}, modelErr
	}

	resp := newExportCharacterResponse(jobID)
	resp.ModelStats = exporter.AggregateModelStats(req.FormatVersion)
	for _, modelFile := range modelPaths {
		info, err := os.Stat(modelFile)
		if err != nil {
			continue
		}
		rel, err := filepath.Rel(outDir, modelFile)
		if err != nil {
			continue
		}
		resp.ExportedModels = append(resp.ExportedModels, exportAssetInfo{Path: filepath.ToSlash(rel), Size: info.Size()})
	}
	for _, tex := range texturePaths {
		info, err := os.Stat(tex)
		if err != nil {
			continue
		}
		rel, err := filepath.Rel(outDir, tex)
		if err != nil {
			continue
		}
		resp.ExportedTextures = append(resp.ExportedTextures, exportAssetInfo{Path: filepath.ToSlash(rel), Size: info.Size()})
	}
	sort.Slice(resp.ExportedTextures, func(i, j int) bool {
		return stringsort.Less(resp.ExportedTextures[i].Path, resp.ExportedTextures[j].Path)
	})
	if !d.Config.IsSharedHosting {
		resp.OutputDirectory = outDir
	}
	summary, _ := json.MarshalIndent(map[string]any{
		"exportedModels":  resp.ExportedModels,
		"outputDirectory": resp.OutputDirectory,
		"versionId":       resp.VersionID,
	}, "", "  ")
	log.Printf("Job finished %s %s", ansi.Yellowf("%.2fs", time.Since(startedAt).Seconds()), ansi.Gray(string(summary)))
	return resp, nil
}

func newExportCharacterResponse(jobID string) exportCharacterResponse {
	return exportCharacterResponse{
		ExportedModels:   []exportAssetInfo{},
		ExportedTextures: []exportAssetInfo{},
		VersionID:        jobID,
	}
}
